add FlashAttentionKwargs and seq_idx to flat collator

garrett361 opened the pull request:
# What does this PR do?

Adds additional, optional return values in `DataCollatorWithFlattening` as needed for padding-free training with particular models.

Relates to #35861 and #35941.

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes # (issue)


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker
- vision models: @amyeroberts, @qubvel
- speech models: @ylacombe, @eustlb
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @zucchini-nlp (visual-language models) or @gante (all others)
- pipelines: @Rocketknight1
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @muellerzr and @SunMarc
- chat templates: @Rocketknight1

Integrations:

- deepspeed: HF Trainer/Accelerate: @muellerzr
- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization (bitsandbytes, autogpt): @SunMarc @MekkCyber

Documentation: @stevhliu

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: See Models above and tag the person corresponding to the modality of the example.
- TensorFlow: @Rocketknight1

 -->

garrett361 committed:
add flash attn kwargs to flattening collator

garrett361 committed:
add return_seq_idx option

vasqu commented:
I'd like to take a look as well when you think you're ready, so gladly ping then :)

vasqu commented:
cc @ArthurZucker

garrett361 committed:
doc string edits

garrett361 commented:
@vasqu could you please take a look when you have time? Thank you.

vasqu reviewed (commented):
Smaller issues/nits overall, I'd be for a warning in case settings might cause issues (RoPE with fa kwarg only)

Otherwise, not on you but I think it would be nice to have equivalent collator tests on torch at least. Usually, all corresponding paths are tested (pt, tf, np).

garrett361 committed:
cleaner max len updates

garrett361 committed:
various fixes

garrett361 commented:
> I think it would be nice to have equivalent collator tests on torch at least

I found it a little surprising that it's all `numpy`, as far as I saw.

garrett361 commented on src/transformers/data/data_collator.py:
```diff
@@ -1786,14 +1786,26 @@ class DataCollatorWithFlattening(DefaultDataCollator):
+    def __init__(
+        self,
+        *args,
+        return_position_ids=True,
+        return_flash_attn_kwargs=False,
```
> maybe someone has a different use case which shouldn't directly cause errors

Yeah, I thought about warnings in cases like that, but I was hesitant because of different requirements for different models.

Like, if a transformer model uses FA but not RoPE, then FA True, pos_ids False make sense. And for a mamba-only model (like mamba2) FA False, pos_ids False, seq_idx True is what you'd use.

garrett361 replied:
So IMO it should be up to the model to ensure it's getting the right inputs it needs and to raise a `ValueError` or similar if an improper combination is passed.

garrett361 replied:
And so actually there should probably be a `return_position_ids: bool =True` arg to cover all cases and not break BC?

vasqu replied:
Fair point, then FA utils should be written in a way that ensures this (which should be a follow up PR to this).

garrett361 replied:
I don't think this can be address at the level of the FA utils, since different models can logically use different valid combinations here.  The FA utils just need to be able to handle the different combos.

It does look like non-trivial `FlashAttentionKwargs` and `position_ids=attention_mask=None` is currently not supported, though. IIUC you'd end up in [this block](https://github.com/huggingface/transformers/blob/60226c6ff3d6bb225782341e58cce5d31f5be1c7/src/transformers/modeling_flash_attention_utils.py?plain=1#L364-L366) with all of your `cu_seq_lens_{q,k}` etc ignored.

vasqu replied:
Couldn't we just detect if fa kwargs were passed (before the if else into the different paths) and handle it then if position_ids is `None`? It might be an error (unintentional path) or warning (no padding path); unsure here. (Or even implementing that path ourselves which seems unwanted)

Imo it's a bit confusing when the original flash attn can handle those args while we can't. As a user it would silently fall through when I'm familiar with fa but not transformers.

garrett361 replied:
> Couldn't we just detect if fa kwargs were passed (before the if else into the different paths) and handle it then if position_ids is None?

Yeah, not sure why this isn't done in the current FA utils. Also found this confusing.

vasqu replied:
Yeah, either way I think some sort of handling is warranted here which should help ease up the bamba checks (hopefully).

garrett361 replied:
Ah, good catch @ArthurZucker , I'll move the `separator_id` back to the second position.

vasqu commented:
Would you be willing to add the other paths or at least pt? Seems like an oversight on the initial PR 👀

garrett361 commented:
> Would you be willing to add the other paths or at least pt? Seems like an oversight on the initial PR

Yep, and I just discovered that the `"pt"` path is broken because `int` entries like those in `cu_seq_lens_{q,k}` turn into `torch.int64`'s whereas FA2 needs them to be `torch.int32`s 😭 

So, this is all going to take some reworking. I'll ping again later.

garrett361 committed:
temp testing code

garrett361 committed:
return int32 seq_idx and FlashAttnKwargs

garrett361 committed:
DataCollatorIntegrationTest impl

garrett361 committed:
fix batch dims and dtypes

garrett361 committed:
fill out remaining collator tests

garrett361 committed:
test name change and fmt

garrett361 committed:
rm unused var

garrett361 committed:
fmt

garrett361 commented:
Alright, made a few changes:
* No batch dimension is expected on any of the `FlashAttentionKwargs`, and all these variables and `seq_idx` must be `int32` rather than the default `int64`. I removed the reliance on the default data collators to achieve this.
* I added a `ModelTesterMixin::test_flash_attention_2_padding_matches_padding_free_with_position_ids_from_flat_collator` which both has an awfully long name and is probably an unnecessarily expensive addition to the test suite.
* Added `tf` and `pt` tests for the flat collator.

Do you have any advice on the test @vasqu ? I wanted a non-trivial test that the collator outputs are in the right format (which is easy to get wrong), but the above seems like overkill.

garrett361 committed:
minor change

garrett361 committed:
fmt

vasqu commented on tests/trainer/test_data_collator.py:
```diff
@@ -1594,9 +1696,68 @@ def test_data_collator_with_flattening(self):
+            batch["input_ids"][0].tolist(), [10, 11, 12, 20, 21, 22, 23, 24, 25, 30, 31, 32, 33, 34, 35, 36]
+        )
+        self.assertNotIn("attention_mask", batch)
+        self.assertIn("position_ids", batch)
```
Kind of a nit but to keep it consistent let's add the value check of positions ids here as well.

garrett361 replied:
Done.

vasqu reviewed (commented):
I think it's overall solid. Some implementations could be simplified imo (e.g. the wrapper at the end of the collator) and left some other smaller comments. 

> No batch dimension is expected on any of the FlashAttentionKwargs, and all these variables and seq_idx must be int32 rather than the default int64. I removed the reliance on the default data collators to achieve this.

Personally, I don't think that the comment about the batch dim provides any real value - would drop it, kinda confused me at first. int32 adjustments are good, only worried about max_seq_len_q/k (should be a simple py in32, but see comments).

> I added a ModelTesterMixin::test_flash_attention_2_padding_matches_padding_free_with_position_ids_from_flat_collator which both has an awfully long name and is probably an unnecessarily expensive addition to the test suite.

Imo, could be added to the previous padding-free test. If you want to keep it as is, I'd suggest a renaming (from flattening collator is rather non-telling).

> Added tf and pt tests for the flat collator.

:partying_face:

vasqu commented on tests/trainer/test_data_collator.py:
```diff
@@ -126,6 +126,31 @@ def test_data_collator_with_padding(self):
         batch = data_collator(features)
         self.assertEqual(batch["input_ids"].shape, torch.Size([2, 8]))
 
+    def test_data_collator_with_flattening(self):
```
The tests across tf, np, and pt should be kept consistent. Atm, the tests differ by values, assertions etc. Would be nice if you could sync them (based on the np tests).

garrett361 commented on tests/trainer/test_data_collator.py:
```diff
@@ -126,6 +126,31 @@ def test_data_collator_with_padding(self):
+        self.assertEqual(batch["max_length_k"].shape, torch.Size([1]))
+        self.assertEqual(batch["cu_seq_lens_q"].dtype, torch.int32)
+        self.assertEqual(batch["cu_seq_lens_k"].dtype, torch.int32)
+        self.assertEqual(batch["max_length_q"].dtype, torch.int32)
+        self.assertEqual(batch["max_length_k"].dtype, torch.int32)
```
I will check, but it seems like it has to be fine if the flash attn test I added passes.

garrett361 committed:
add missing pos_ids check

garrett361 committed:
consistent {np,pt,tf} tests

garrett361 committed:
split pt tests into 3, like np/tf tests

garrett361 committed:
mv comment, rename fa test

garrett361 commented on src/transformers/data/data_collator.py:
```diff
@@ -1813,12 +1826,58 @@ def __call__(self, features, return_tensors=None, separator_id=None):
+        if self.return_flash_attn_kwargs:
+            ret["cu_seq_lens_q"] = data_cls(ret["cu_seq_lens_q"], dtype=dtype_32)
+            ret["cu_seq_lens_k"] = data_cls(ret["cu_seq_lens_k"], dtype=dtype_32)
+            ret["max_length_q"] = data_cls(ret["max_length_q"], dtype=dtype_32)
+            ret["max_length_k"] = data_cls(ret["max_length_k"], dtype=dtype_32)
```
Looks like it is typed as a simple `int` in the FA2 library, but a `torch.Tensor` is apparently also accepted just fine.

Also note that HF's `prepare_fa2_from_position_ids` utility claims to return `max_seqlen_in_batch_{q,k}` as `int`s, but it actually returns a `torch.Tensor`
https://github.com/huggingface/transformers/blob/7f5077e53682ca855afc826162b204ebf809f1f9/src/transformers/modeling_flash_attention_utils.py?plain=1#L181

Do you want me to change the collator behavior to `int`s?

vasqu replied:
Ah that's interesting, was used to the fa typing. If it works, it's fine for now, so keep it as is. That also keeps things consistent with the other function(s) (even if the typing is wrong atm).

garrett361 committed:
remove batch dim comment

garrett361 committed:
simply wrapping

garrett361 commented on src/transformers/data/data_collator.py:
```diff
@@ -1813,12 +1826,58 @@ def __call__(self, features, return_tensors=None, separator_id=None):
+            raise ValueError(f'return_tensors must be one of ("pt", "np", "tf"), not {return_tensors=}')
+
+        ret["input_ids"] = data_cls(ret["input_ids"], dtype=dtype_64)[None]
+        ret["labels"] = data_cls(ret["labels"], dtype=dtype_64)[None]
```
Cleaned up.

garrett361 commented on src/transformers/data/data_collator.py:
```diff
@@ -1813,12 +1826,58 @@ def __call__(self, features, return_tensors=None, separator_id=None):
+                ret["seq_idx"] += [seq_idx for _ in range(len(input_ids))]
+            if self.return_flash_attn_kwargs:
+                len_input_ids = len(input_ids)
+                ret["cu_seq_lens_q"].append(ret["cu_seq_lens_q"][-1] + len_input_ids)
+                ret["cu_seq_lens_k"].append(ret["cu_seq_lens_k"][-1] + len_input_ids)
+                ret["max_length_q"][0] = max(ret["max_length_q"][0], len_input_ids)
+                ret["max_length_k"][0] = max(ret["max_length_k"][0], len_input_ids)
```
Do you mean just compute `cu_seq_lens_{q,k}` and `max_length_{q,k}` once at the end, rather than in a loop? That seems better, will do. LMK if I misunderstood.

vasqu replied:
Had in mind to only compute one of the since both share the same value(s). E.g. we only calculate max_length_q. Afterwards we copy that value to max_length_k.

But honestly, the approach of calculating it afterwards is even better! Good thinking

garrett361 replied:
Guess it's a tradeoff of having a single loop over `features`, but paying the cost of more `append` and `max` calls, versus more loops over `features`.  Not sure which is better.

vasqu replied:
Hmm, that's also true. I don't think it's very important for now, so let's keep it simple and let it be for now. Wdyt?

garrett361 replied:
Yeah, started rewriting it and it didn't seem any better code wise, so leaving it.

garrett361 replied:
Done.

vasqu reviewed (approved):
LGTM! cc @ArthurZucker for core maintainer review

vasqu commented:
Looks like you'll need to run `make style` for the ci to be happy. I think the PR is overall ready.

garrett361 committed:
compute cu_seq_len/max_length once

garrett361 committed:
fmt

garrett361 commented:
@ArthurZucker please let me know if I can answer any further questions

garrett361 commented:
@vasqu any advice here?

vasqu commented:
@garrett361 sorry don't have anything to add. Arthur is quite busy so notifications can easily get lost at times, I'm gonna cc again.

vasqu commented:
cc core maintainer review @ArthurZucker @Cyrilvallez

Cyrilvallez commented on src/transformers/data/data_collator.py:
```diff
@@ -1902,12 +1921,53 @@ def __call__(self, features, return_tensors=None, separator_id=None):
         ret = {"input_ids": [], "labels": []}
         if self.return_position_ids:
             ret.update({"position_ids": []})
-        for idx in range(0, len(features)):
-            ret["input_ids"] += features[idx]["input_ids"]
+        if self.return_seq_idx:
+            ret.update({"seq_idx": []})
```
`ret` is quite a bad variable name -> let's take advantage of this PR to switch to `output` or something along those lines! 🤗

garrett361 replied:
`ret -> batch`

ArthurZucker commented on src/transformers/data/data_collator.py:
```diff
@@ -1902,12 +1921,53 @@ def __call__(self, features, return_tensors=None, separator_id=None):
+            ret["max_length_q"] = ret["max_length_k"] = [max_length]
+
+        # FlashAttentionKwargs and seq_idx are expected to be int32s.
+        if return_tensors == "pt":
+            import torch
+
+            data_cls = torch.tensor
+            dtype_64 = torch.int64
+            dtype_32 = torch.int32
+        elif return_tensors == "np":
+            data_cls = np.array
+            dtype_64 = np.int64
+            dtype_32 = np.int32
+        elif return_tensors == "tf":
+            import tensorflow as tf
+
+            data_cls = tf.convert_to_tensor
+            dtype_64 = tf.int64
+            dtype_32 = tf.int32
+        else:
+            raise ValueError(f'return_tensors must be one of ("pt", "np", "tf"), not {return_tensors=}')
```
Happy to go with torch only

ArthurZucker replied:
let's just fix this and we can merge

Cyrilvallez replied:
Agreed, tensorflow models will never support it anyway!

garrett361 replied:
Yeah, was just going for completeness since it was easy. Will delete the TF and associated test code.

ArthurZucker reviewed (approved):
Thanks for your PR, sorry that it got lost 😓 merging 🤗

Cyrilvallez commented on tests/trainer/test_data_collator.py:
```diff
@@ -126,6 +126,83 @@ def test_data_collator_with_padding(self):
+        self.assertEqual(
+            batch["input_ids"][0].tolist(), [10, 11, 12, 20, 21, 22, 23, 24, 25, 30, 31, 32, 33, 34, 35, 36]
+        )
+        self.assertNotIn("attention_mask", batch)
+        self.assertNotIn("cu_seq_lens_k", batch)
+        self.assertNotIn("cu_seq_lens_q", batch)
+        self.assertNotIn("max_length_k", batch)
+        self.assertNotIn("max_length_q", batch)
+        self.assertNotIn("seq_idx", batch)
+        self.assertIn("position_ids", batch)
```
Here (and similarly afterwards) it would be clearer to match a list instead of individuals checks!

Cyrilvallez reviewed (commented):
Just a few nits on top of Arthur's review as I was looking at it at the same time!

garrett361 commented:
> Thanks for your PR, sorry that it got lost

No worries, I know you're super busy. Will make these changes and ping.

garrett361 committed:
remove tf code

garrett361 committed:
rm warning

garrett361 committed:
move separator_id back to 2nd pos

garrett361 committed:
use cleaner lists in tests

garrett361 committed:
ret -> batch

garrett361 committed:
fmt

garrett361 commented on src/transformers/data/data_collator.py:
```diff
@@ -1888,10 +1890,27 @@ class DataCollatorWithFlattening(DefaultDataCollator):
+        warnings.warn(
+            "Using `DataCollatorWithFlattening` will flatten the entire mini batch into single long sequence."
+            "Make sure your attention computation is able to handle it!"
+        )
```
Removed.

garrett361 committed:
attr ordering

garrett361 commented:
The failing test is due to infra issues:
```
Couldn't find 'hf-internal-testing/librispeech_asr_dummy'
```

Believe I made all the requested changes. CC @ArthurZucker @Cyrilvallez

vasqu reviewed (commented):
Seems like the torch version for max_seqlen_(q/k) causes compile issues. See #37206

I know we've discussed this before and let it be but I think it's warranted to change it up for now. @garrett361

garrett361 commented:
Ah, yeah, if that's causing compile problems then let me fix it.

garrett361 committed:
use py ints for max_length_{k,q}

garrett361 commented:
Alright @vasqu `max_len_{q,k}` are returned as simple py ints now.

garrett361 commented:
Hi @Cyrilvallez and @ArthurZucker , a reminder about this one, please.

Cyrilvallez reviewed (approved):
Merging! Thanks a lot for the great PR! 🤗 Super sorry we took this long!

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_36456). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

garrett361 commented:
Thanks @Cyrilvallez !
